src/shim.js

```diff
@@ -4,7 +4,89 @@
 
 "use strict";
 
-class UrlbarProvider {}
+class UrlbarProvider {
+  constructor(name) {
+    this.onBehaviorRequested = this.onBehaviorRequested.bind(this);
+    this.onResultsRequested = this.onResultsRequested.bind(this);
+    this.onViewUpdateRequested = this.onViewUpdateRequested.bind(this);
+    this.onResultPicked = this.onResultPicked.bind(this);
+    // this.onEngagement = this.onEngagement.bind(this);
+    this.onQueryCanceled = this.onQueryCanceled.bind(this);
+
+    browser.urlbar.onBehaviorRequested.addListener(
+      this.onBehaviorRequested,
+      this.name
+    );
+    browser.urlbar.onResultsRequested.addListener(
+      this.onResultsRequested,
+      this.name
+    );
+    browser.experiments.urlbar.onViewUpdateRequested.addListener(
+      this.onViewUpdateRequested,
+      this.name
+    );
+    browser.urlbar.onResultPicked.addListener(this.onResultPicked, this.name);
+    // browser.urlbar.onEngagement.addListener(this.onEngagement, this.name);
+    browser.urlbar.onQueryCanceled.addListener(this.onQueryCanceled, this.name);
+  }
+
+  removeListeners() {
+    browser.urlbar.onBehaviorRequested.removeListener(
+      this.onBehaviorRequested,
+      this.name
+    );
+    browser.urlbar.onResultsRequested.removeListener(
+      this.onResultsRequested,
+      this.name
+    );
+    browser.experiments.urlbar.onViewUpdateRequested.removeListener(
+      this.onViewUpdateRequested,
+      this.name
+    );
+    browser.urlbar.onResultPicked.removeListener(
+      this.onResultPicked,
+      this.name
+    );
+    // browser.urlbar.onEngagement.removeListener(this.onEngagement, this.name);
+    browser.urlbar.onQueryCanceled.removeListener(
+      this.onQueryCanceled,
+      this.name
+    );
+  }
+
+  async onBehaviorRequested(query) {
+    if (!(await this.isActive(query))) {
+      return "inactive";
+    }
+    if ((await this.getPriority(query)) > 0) {
+      return "restricting";
+    }
+    return "active";
+  }
+
+  async onResultsRequested(query) {
+    let results = [];
+    await this.startQuery(query, (p, result) => results.push(result));
+    return results;
+  }
+
+  async onViewUpdateRequested(payload) {
+    return this.getViewUpdate({ payload });
+  }
+
+  async onResultPicked(payload) {
+    await this.pickResult({ payload });
+  }
+
+  //     async onEngagement(state) {
+  //       let isPrivate = false; //XXXadw
```

Comment:
Left in?

Reply:
I'd assumed that you'd left it in as you might want it, but I've removed it for now.